backend/store/stacktraces.go

```diff
@@ -0,0 +1,258 @@
+		commitSha, _, err := gitHubClient.GetLatestCommitHash(ctx, gitHubRepoPath)
+		if err != nil {
+			return nil, err
+		}
+		return &commitSha, nil
+	})
+}
+
+func (store *Store) EnhanceTraceWithGitHub(ctx context.Context, trace *privateModel.ErrorTrace, service *model.Service, serviceVersion string, ignoredFiles []string, gitHubClient github.ClientInterface) (*privateModel.ErrorTrace, error) {
+	if trace.FileName == nil || trace.LineNumber == nil {
+		return trace, fmt.Errorf("Cannot enhance trace with GitHub with invalid values: %+v", trace)
+	}
+
+	fileName := store.GitHubFilePath(ctx, *trace.FileName, service.BuildPrefix, service.GithubPrefix)
+	lineNumber := trace.LineNumber
+
+	for _, fileExpr := range ignoredFiles {
+		if regexp.MustCompile(fileExpr).MatchString(fileName) {
+			return trace, nil
+		}
+	}
+
+	gitHubFileBytes, err := store.storageClient.ReadGitHubFile(ctx, *service.GithubRepoPath, fileName, serviceVersion)
+
+	if err != nil || gitHubFileBytes == nil {
+		encodedFileContent, err := store.FetchFileFromGitHub(ctx, trace, service, fileName, serviceVersion, gitHubClient)
+		if err != nil {
+			return trace, err
+		}
+
+		gitHubFileBytes = []byte(*encodedFileContent)
+
+		_, err = store.storageClient.PushGitHubFile(ctx, *service.GithubRepoPath, fileName, serviceVersion, gitHubFileBytes)
+		if err != nil {
+			log.WithContext(ctx).Error(err)
+		}
+	}
+
+	gitHubFileString := string(gitHubFileBytes)
+	rawDecodedText, err := base64.StdEncoding.DecodeString(gitHubFileString)
+	if err != nil || len(rawDecodedText) == 0 {
+		return trace, err
+	}
+
+	lines := strings.Split(string(rawDecodedText), "\n")
+	lineContent, beforeContent, afterContent, err := store.ExpandedStackTrace(ctx, lines, *lineNumber)
+	if err != nil {
+		return trace, err
+	}
+
+	newStackTraceInput := privateModel.ErrorTrace{
+		FileName:                   trace.FileName,
+		LineNumber:                 trace.LineNumber,
+		FunctionName:               trace.FunctionName,
+		Error:                      trace.Error,
+		SourceMappingErrorMetadata: trace.SourceMappingErrorMetadata,
+		LineContent:                lineContent,
+		LinesBefore:                beforeContent,
+		LinesAfter:                 afterContent,
+	}
+	return &newStackTraceInput, nil
+}
+
+func (store *Store) GitHubEnhancedStakeTrace(ctx context.Context, stackTrace []*privateModel.ErrorTrace, projectID int, errorObj *model.ErrorObject, serviceName string, serviceVersion string) ([]*privateModel.ErrorTrace, error) {
+	if serviceName == "" {
+		return nil, nil
+	}
+
+	service, err := store.FindService(ctx, projectID, serviceName)
+	if err != nil || service == nil || service.GithubRepoPath == nil || service.Status != "healthy" {
+		return nil, err
+	}
+
+	workspace := &model.Workspace{}
+	if err := store.db.Model(&model.Workspace{}).Joins("LEFT JOIN projects ON projects.workspace_id = workspaces.id").Where("projects.id = ?", projectID).Take(&workspace).Error; err != nil {
```

Comment:
can also `.Preload("Projects")` instead of manually joining

Reply:
Is there a good place I can get the `workspace_id` here? Basically trying to find the workspace, but the only thing I have to filter on is the `project_id`?

Reply:
I was able to refactor this to avoid needing to hit the db, and just pass in the workspace and project as arguments. The method is getting a little large with 7 arguments, but let me know if this is a concern
```
func (store *Store) EnhancedStackTrace(ctx context.Context, stackTrace string, workspace *model.Workspace, project *model.Project, errorObj *model.ErrorObject, serviceName string, serviceVersion string) (*string, []*privateModel.ErrorTrace, error)
```